homework/ba-nguyen/lesson-7.2/javascript-3.md

```diff
@@ -0,0 +1,249 @@
+**Question 1**: Write a **function printNumbers(from, to)** that outputs a number every second, starting from from and ending with to. Make two variants of the solution.
+
+- Using setInterval
+
+```typescript
+function printNumbers(from: number, to: number) {
+  let timeout: any;
+
+  timeout = setInterval(() => {
+    if (from <= to) {
+      console.log(from);
+    }
+
+    if (from === to) {
+      clearInterval(timeout);
+    }
+
+    from++;
+  }, 1000);
+}
+```
+
+- Using nested setTimeout
+
+```typescript
+function printNumbers(from: number, to: number) {
+  if (from > to) return;
+
+  setTimeout(() => {
+    console.log(from);
+    printNumbers(from + 1, to);
+  }, 1000);
+}
+```
+
+**Question 2**: Use the **reduce** method in combination with the **concat** method to “flatten” an array of arrays into a single array that has all the elements of the original arrays.
+
+```typescript
+const arrays = [[1, 2, 3], [4, 5], [6]];
+→ [1, 2, 3, 4, 5, 6]
+
+function flatten(arrays: any[], depth: number = Infinity): any[] {
+  if (!Array.isArray(arrays)) {
+    return arrays;
+  }
+
+  if (depth > 0) {
+    return arrays.reduce(
+      (prevArr, curVal) => prevArr.concat(flatten(curVal, depth - 1)),
+      []
+    );
+  }
+
+  return arrays.slice();
+}
+```
+
+**Question 3**: Write a higher-order function loop that provides something like a for loop statement. It takes a value, a test function, an update function, and a body function. Each iteration, it first runs the test function on the current loop value and stops if that returns false. Then it calls the body function, giving it the current value. Finally, it calls the update function to create a new value and starts from the beginning. When defining the function, you can use a regular loop to do the actual looping.
+
+loop(3, n => n > 0, n => n - 1, console.log);
+// → 3
+// → 2
+// → 1
+
+```typescript
+type TestFuncType = (n: number) => boolean;
+type UpdateFuncType = (n: number) => number;
+type BodyFuncType = (n: number) => void;
+
+function loop(
+  value: number,
+  testFunc: TestFuncType,
+  updateFunc: UpdateFuncType,
+  bodyFunc: BodyFuncType
+) {
+  while (testFunc(value)) {
```

Comment:
this question require use higher-order function

Reply:
Can you give me a hint? i still don't your mean

Reply:
higher-order function need return function
```js
function loop(value, test, update, execute) {
  if (!test(value)) return;
  execute(value);
  return loop(update(value), test, update, execute);
}

loop(3, n => n > 0, n => n - 1, console.log);
```